src/ripple/app/tests/Env.cpp

```diff
@@ -0,0 +1,217 @@
+AccountInfo::seq() const
+{
+    return root_->getFieldU32(sfSequence);
+}
+
+std::uint32_t
+AccountInfo::flags() const
+{
+    return root_->getFieldU32(sfFlags);
+}
+
+//------------------------------------------------------------------------------
+
+Env::Env (beast::unit_test::suite& test_)
+    : test(test_)
+    , master("master", generateKeysFromSeed(
+        KeyType::secp256k1, RippleAddress::createSeedGeneric(
+            "masterpassphrase")))
+{
+    memoize(master);
+    initializePathfinding();
+    ledger = std::make_shared<Ledger>(
+        master.pk(), SYSTEM_CURRENCY_START);
+}
+
+void
+Env::memoize (Account const& account)
+{
+    map_.emplace(account.id(), account);
+}
+
+Account const&
+Env::lookup (std::string const& base58ID) const
+{
+    RippleAddress ra;
+    if (! ra.setAccountID(base58ID))
+        throw std::runtime_error(
+            "Env::lookup: invalid account ID");
+    return lookup(ra.getAccountID());
+}
+
+Account const&
+Env::lookup (ripple::Account const& id) const
+{
+    auto const iter = map_.find(id);
+    if (iter == map_.end())
+        throw std::runtime_error(
+            "Env::lookup:: unknown account ID");
+    return iter->second;
+}
+
+void
+Env::fund (STAmount const& amount,
+    Account const& account)
+{
+    using namespace jtx;
+    memoize(account);
+    apply(pay(master, account, amount),
+        seq(jtx::autofill),
+            fee(jtx::autofill),
+                sig(jtx::autofill));
+}
+
+void
+Env::trust (STAmount const& amount,
+    Account const& account)
+{
+    using namespace jtx;
+    apply(jtx::trust(account, amount),
+        seq(jtx::autofill),
+            fee(jtx::autofill),
+                sig(jtx::autofill));
+}
+
+void
+Env::apply (JTx const& tx)
+{
+    boost::optional<STTx> stx;
+    {
+        // The parse must succeed, since we
+        // generated the JSON ourselves.
+        boost::optional<STObject> st;
+        try
+        {
+            st = jtx::parse(tx.jv);
+        }
+        catch(jtx::parse_error const&)
+        {
+            test.log << pretty(tx.jv);
+            throw;
+        }
+
+        try
+        {
+            stx.emplace(std::move(*st));
+        }
+        catch(...)
+        {
+        }
+    }
+
+    TER ter;
+    bool didApply;
+    if (stx)
+    {
+        TransactionEngine txe (ledger, multisign);
+        std::tie(ter, didApply) = txe.applyTransaction(
+            *stx, tapOPEN_LEDGER |
+                (true ? tapNONE : tapNO_CHECK_SIGN));
+    }
+    else
+    {
+        // Convert the exception into a TER so that
+        // callers can expect it using ter(temMALFORMED)
+        ter = temMALFORMED;
+        didApply = false;
+    }
+    if (! test.expect(ter == tx.ter,
+        "apply: " + transToken(ter) +
+            " (" + transHuman(ter) + ")"))
+        test.log << pretty(tx.jv);
+}
+
+void
+Env::autofill (JTx& jt)
+{
+    auto& jv = jt.jv;
+    auto const should = [](boost::tribool v, bool b)
+    {
+        if (v)
```

Comment:
`tribool` is converted to a bool as: `true->true` otherwise false. Many of your uses of tribool can be simplified to:
```
if (boost::indeterminate(tribool))
   return some_bool;
else
  return bool(tribool);
}
```